scvid/data/dadc_dataset.py

```diff
@@ -14,11 +16,161 @@ def __init__(self, dadc: DistributedAnnDataCollection) -> None:
     def __len__(self) -> int:
         return len(self.dadc)
 
-    def __getitem__(self, idx: int) -> Dict[str, Tensor]:
+    def __getitem__(self, idx: int) -> Dict[str, np.ndarray]:
         """Return gene counts for a cell at idx."""
         X = self.dadc[idx].X
 
         data = {}
         data["X"] = X.toarray() if issparse(X) else X
 
         return data
+
+
+class IterableDistributedAnnDataCollectionDataset(IterableDataset):
+    r"""
+    Iterable DistributedAnnDataCollectionDataset.
+
+    Args:
+        dadc: DistributedAnnDataCollection from which to load the data.
+        batch_size: how many samples per batch to load. Default: ``1``.
+        shuffle: set to ``True`` to have the data reshuffled
+            at every epoch. Default: ``False``.
+        seed: random seed used to shuffle the sampler if :attr:`shuffle=True`. Default: ``0``.
+        test_mode: If ``True`` enables tracking of cache and worker informations.
+    """
+
+    def __init__(
+        self,
+        dadc: DistributedAnnDataCollection,
+        batch_size: int = 1,
+        shuffle: bool = False,
+        seed: int = 0,
+        test_mode: bool = False,
+    ) -> None:
+        self.dadc = dadc
+        self.batch_size = batch_size
+        self.shuffle = shuffle
+        self.seed = seed
+        self.epoch = 0
+        self.test_mode = test_mode
+
+    def __len__(self) -> int:
+        return len(self.dadc)
+
+    def set_epoch(self, epoch: int) -> None:
+        r"""
+        Sets the epoch for the iterator. When :attr:`shuffle=True`, this ensures all replicas
+        use a different random ordering for each epoch.
+        """
+        self.epoch = epoch
+
+    def __getitem__(self, idx: Union[int, List[int]]) -> Dict[str, np.ndarray]:
+        """Return gene counts for cells at idx."""
+        X = self.dadc[idx].X
+
+        data = {}
+        data["X"] = X.toarray() if issparse(X) else X
+
+        # for testing purposes
+        if self.test_mode:
+            worker_info = get_worker_info()
+            if worker_info is not None:
+                data["worker_id"] = np.array([worker_info.id])
+            data["miss_count"] = np.array([self.dadc.cache.miss_count])
+
+        return data
+
+    def __iter__(self):
```

Comment:
Please extend the Docstring: that we use `torch.utils.data.get_worker_info()` to determine the invocation context. If single worker, then we iterate over the entire datasets. Otherwise, we iterate over a subset of cells in a manner that minimizes the overlap between the data chunks loaded by each worker.

Reply:
@ordabayevy solid work!

What's the plan for multi-GPU? One approach is to create separate DADCs for each worker based on their rank. Another approach (perhaps cleaner) would be to get the world size and node rank from `torch.distributed` and treat `work_size x num_workers` as `total_workers`. The same logic that you've implemented should work simply by replacing `num_workers` with `total_workers`.

I would also suggest factoring out the code that generate `indices` out of `__iter__`. Then you can make an issue to make a smarter assignment strategy where you balance the number of cells given to each worker using a greedy algorithm (this is low priority for our current use case).

Reply:
Thanks for reviewing @mbabadi, addressed your comments. I have an implemented version for the multi-GPU case here https://github.com/cellarium-ai/scvi-distributed/pull/21/files.
Working on tests at the moment :) We can discuss it tomorrow at our meeting.